fakestorage/upload.go

```diff
@@ -150,17 +150,54 @@ func (s *Server) insertFormObject(r *http.Request) xmlResponse {
 func (s *Server) checkUploadPreconditions(r *http.Request, bucketName string, objectName string) *jsonResponse {
 	ifGenerationMatch := r.URL.Query().Get("ifGenerationMatch")
 
-	if ifGenerationMatch == "0" {
-		if _, err := s.backend.GetObject(bucketName, objectName); err == nil {
+	if ifGenerationMatch != "" {
+		gen, err := strconv.ParseInt(ifGenerationMatch, 10, 64)
+		if err != nil {
+			return &jsonResponse{
+				status:       http.StatusBadRequest,
+				errorMessage: err.Error(),
+			}
+		}
+		_, err = s.backend.GetObjectWithGeneration(bucketName, objectName, gen)
+		if gen == 0 {
+			if err == nil {
```

Comment:
Shouldn't it be `if err != nil` as it used to be before? With this change, [my code](https://github.com/orsinium-labs/dimutex/blob/master/dimutex/_gcs.py#L117-L120) started to get `OK` status where a `Precondition Failed` status is expected and I'm trying to find out why. This line is my primary suspect.

Reply:
Hmm, I don't think so. ifGenerationMatch=0 means that the object must not exist. `err == nil` means the object exists, so that's why this path fails with `precondition failed`.

Do you have a reproducer or a request dump?

Reply:
I've also checked. You're right, changing this line doesn't help.

I linked the code of the library I use. If I don't find the cause myself, I'll open an issue with steps to reproduce. Thank you for a quick response.

Reply:
Oh, I guess the difference from before is that we're invoking `GetObjectWithGeneration` instead of `GetObject`, and that fails with the filesystem backend, because it doesn't support generations currently. We should really implement generations on that backend.

Is using the memory backend an option for you? Or are you dealing with large objects? I guess we could special-case ifGenerationMatch=0 to fix the breaking change while the filesystem backend doesn't support it. Any thoughts?

Reply:
Confirmed, using `-backend memory` fixes the bug. For me, in-memory backend should work but it's good to have it fixed. I'll create an issue for it, I already made a small Python script to reproduce the bug.

Reply:
Ah, I see #744. Thank you. I'll just comment there.

Reply:
Yeah implementing generation support in the filesystem backend is going to take some work, but I'm working on a quickfix for when ifGeneration=0 to keep it backwards compatible. I should have a PR very soon.